src/profile/webid.ts

```diff
@@ -0,0 +1,106 @@
+ * to be a Solid resource. That is why, in order to expose a Solid-enabled part
+ * of their profile, some WebID profile documents link to a personal profile
+ * documents, which may be a Solid resource.
+ *
+ * @param webIdDocument data from the WebID profile document
+ * @param options Optional parameter `options.fetch`: An alternative `fetch`
+ *  function to make the HTTP request, compatible with the browser-native
+ *  [fetch API](https://developer.mozilla.org/docs/Web/API/WindowOrWorkerGlobalScope/fetch#parameters).
+ * @returns Promise resolving to an array of [[SolidDataset]], each corresponding
+ *  to a personal profile document discoverable from the WebID Profile Document.
+ *  If none are found, the WebID profile document itself is returned.
+ */
+export async function getWebIdProfileAll(
+  webIdDocument: SolidDataset & WithResourceInfo,
+  options: Partial<
+    typeof internal_defaultFetchOptions
+  > = internal_defaultFetchOptions
+): Promise<SolidDataset[]> {
+  const foundProfiles = await Promise.all(
+    getThingAll(webIdDocument)
+      .filter((thing) => getIriAll(thing, foaf.primaryTopic).length > 0)
+      // This assumes that getSourceIri returns the IRI which has been fetched to
+      // get the profile document.
+      .filter((thing) => asIri(thing) !== getSourceIri(webIdDocument))
+      .map(
+        async (primaryThing) =>
+          await getSolidDataset(asIri(primaryThing), { fetch: options.fetch })
+      )
+  );
+  return foundProfiles.length > 0 ? foundProfiles : [webIdDocument];
+}
+
+/**
+ * Get a personal profiles discoverable from a WebID profile document.
+ *
+ * A WebID profile document may be any RDF resource on the Web, it doesn't have
+ * to be a Solid resource. That is why, in order to expose a Solid-enabled part
+ * of their profile, some WebID profile documents link to a personal profile
+ * documents, which may be a Solid resource.
+ *
+ * @param webIdDocument data from the WebID profile document
+ * @param options Optional parameter `options.fetch`: An alternative `fetch`
+ *  function to make the HTTP request, compatible with the browser-native
+ *  [fetch API](https://developer.mozilla.org/docs/Web/API/WindowOrWorkerGlobalScope/fetch#parameters).
+ * @returns Promise resolving to a [[SolidDataset]] corresponding to a personal
+ *  profile document discoverable from the WebID Profile Document. If several exist,
+ *  one is arbitrarily chosen.
+ *  If none are found, the WebID profile document itself is returned.
+ */
+export async function getWebIdProfile(
+  webIdDocument: SolidDataset & WithResourceInfo,
+  options: Partial<
+    typeof internal_defaultFetchOptions
+  > = internal_defaultFetchOptions
+): Promise<SolidDataset> {
+  // Arbitrarily pick a subject of a triple <s, foaf:primaryTopic, o>
+  const [primaryThing] = getThingAll(webIdDocument)
+    .filter((thing) => getIriAll(thing, foaf.primaryTopic).length > 0)
```

Comment:
We need to catter for the inverse property `foaf:isPrimaryTopicOf`.

Reply:
```suggestion
  const [primaryThing] = getThingAll(webIdDocument)
    .filter((thing) => getIriAll(thing, foaf.primaryTopic).length > 0 || getIriAll(thing, foaf.isPrimaryTopicOf).length > 0)
```
Something like that but taking predicate object instead of subject predicate. Not sure what that function is in the SDK.

Reply:
This should now be resolved.